### DISCUSSION TITLE
Z2M resends old state before sending new state

### DISCUSSION THREAD

**Discussion by Syntaxrabbit** [Koenkk/zigbee2mqtt#12806] [General]
Good morning!

Unfortunately I did not find anything related with the search funtionality, so I hope some experts here can shed some light on my issue.

I connected Z2M with openhab which works great most of the time.
Now, I have a new use case which works like that:

- Wall switch triggers an item in openhab (ON/OFF)
- openhab rule triggers Z2M binding accordingly
- LED bulb gets switched on or off according to wall switch
- LED bulb may get triggered by any other device
- Z2M sends back state to openhab to update the item (for the correct bulb state in user interface, but only if old state != new state)

Now my problem is, that Z2M first sends the old state again and immediately the new state via MQTT:

`MQTT publish: topic 'zigbee2mqtt/LED-StehlampeWohnzimmer', payload '{"brightness":144,"color":{"h":112,"hue":239.1,"saturation":99,"x":0.1405,"y":0.0547},"color_mode":"hs","color_temp":526,"last_seen":1655104079231,"linkquality":120,"state":"OFF","update":{"state":"idle"},"update_available":false}'`

`MQTT publish: topic 'zigbee2mqtt/LED-StehlampeWohnzimmer', payload '{"brightness":144,"color":{"h":112,"hue":239.1,"saturation":99,"x":0.1405,"y":0.0547},"color_mode":"hs","color_temp":526,"last_seen":1655104079231,"linkquality":120,"state":"ON","update":{"state":"idle"},"update_available":false}'`

This leads to a dead lock:

- I press the switch
- openhab sends "on" to Z2M
- Z2M sends "off" to openhab
- Z2M sends "on" to openhab
- in between openhab resends "off" to Z2M because the status changed
- on-off-loop and both bulb and switch start blinking

My question is, if this Z2M behaviour is intended, or if it is a bug, or maybe it can be disabled somewhere in the settings?

Thank you all for your help.

**Comment by joconcepts** [comment]
I got the same issue with some OSRAM power outlet:

```
Debug 2022-08-31 20:31:11Received MQTT message on 'zigbee2mqtt/Flur Schranklampe/set' with data '{"state":"OFF"}'
Debug 2022-08-31 20:31:11Publishing 'set' 'state' to 'Flur Schranklampe'
Info 2022-08-31 20:31:11MQTT publish: topic 'zigbee2mqtt/Flur Schranklampe', payload '{"last_seen":"2022-08-31T18:31:11.506Z","linkquality":87,"state":"ON","update":{"state":"idle"},"update_available":false}'
Info 2022-08-31 20:31:11MQTT publish: topic 'zigbee2mqtt/Flur Schranklampe', payload '{"last_seen":"2022-08-31T18:31:11.506Z","linkquality":87,"state":"OFF","update":{"state":"idle"},"update_available":false}'
Debug 2022-08-31 20:31:11Received Zigbee message from 'Flur Schranklampe', type 'attributeReport', cluster 'genOnOff', data '{"onOff":0}' from endpoint 1 with groupID 0
Info 2022-08-31 20:31:11MQTT publish: topic 'zigbee2mqtt/Flur Schranklampe', payload '{"last_seen":"2022-08-31T18:31:11.518Z","linkquality":87,"state":"OFF","update":{"state":"idle"},"update_available":false}'
Debug 2022-08-31 20:31:12Received MQTT message on 'zigbee2mqtt/Flur Schranklampe/set' with data '{"state":"ON"}'
Debug 2022-08-31 20:31:12Publishing 'set' 'state' to 'Flur Schranklampe'
Info 2022-08-31 20:31:12MQTT publish: topic 'zigbee2mqtt/Flur Schranklampe', payload '{"last_seen":"2022-08-31T18:31:12.759Z","linkquality":87,"state":"OFF","update":{"state":"idle"},"update_available":false}'
Info 2022-08-31 20:31:12MQTT publish: topic 'zigbee2mqtt/Flur Schranklampe', payload '{"last_seen":"2022-08-31T18:31:12.759Z","linkquality":87,"state":"ON","update":{"state":"idle"},"update_available":false}'
```

When sending state "OFF" it first publishes "ON" then "OFF". Homebridge does not like that and shows the wrong state.

Modell: OSRAM AC10691

**Comment by Syntaxrabbit** [comment]
Any news on this? It still seems to be an issue... I solved it in Openhab by creating a rule which triggers in both directions by waiting a whole second, which is kind of extremly ugly. (But it works for now...)

**Comment by nobbi1991** [comment]
I have the same issue.
Is there some Zigbee2MQTT setting which can be changed to avoid sending the previous state?

**Comment by plopes9000** [comment]
Came across the same issue. Turns out the cause of this behavior is the setting below which can be set in the zigbee2mqtt UI:
  advanced:
    # Optional: state caching, MQTT message payload will contain all attributes, not only changed ones.
    # Has to be true when integrating via Home Assistant (default: true)
    cache_state: true

Set if to false and the old state will no longer be reported.

https://www.zigbee2mqtt.io/guide/configuration/mqtt.html#server-connection // MQTT behavior
Surprisingly the setting is not in the MQTT tab, rather in the general, Settings, Advanced, after "Adapter delay"

**Comment by Syntaxrabbit** [comment]
Maybe I am blind, but where exactly is this setting? I cannot find it on the user interface. Or is it inside a configuration file?

**Reply by plopes9000** [reply]
I've updated my first comment, see above - this way someone else does not have to follow the thread to get the answer.
